lib/block_view_lockups.go

```diff
@@ -0,0 +1,1784 @@
+		if txMeta.LockupAmountBaseUnits.Gt(transactorBalanceNanos256) {
+			return 0, 0, nil,
+				errors.Wrap(RuleErrorCoinLockupInsufficientCoins, "_connectCoinLockup")
+		}
+
+		// We store the previous transactor balance entry in the event we need to revert the transaction.
+		prevTransactorBalanceEntry = transactorBalanceEntry
+
+		// Spend the transactor's DAO coin balance.
+		transactorBalanceEntry.BalanceNanos =
+			*uint256.NewInt().Sub(&transactorBalanceEntry.BalanceNanos, txMeta.LockupAmountBaseUnits)
+		bav._setDAOCoinBalanceEntryMappings(transactorBalanceEntry)
+	}
+
+	// By now we know the transaction to be valid. We now source yield information from either
+	// the profile's yield curve or the raw DeSo yield curve. Because there's some choice in how
+	// to determine the yield when the lockup duration falls between two profile specified yield curve
+	// points, we return here the two local points and choose/interpolate between them below.
+	leftYieldCurvePoint, rightYieldCurvePoint := bav.GetLocalYieldCurvePoints(profilePKID, lockupDurationNanoSeconds)
+
+	// Here we interpolate (choose) the yield between the two returned local yield curve points.
+	//
+	// If we fall between two points, we choose the left yield curve point (i.e. the one with lesser lockup duration).
+	// The transactor earns yield only for the lockup duration specified by the left yield curve point but will
+	// be unable to unlock the coins until the transaction specified lockup duration expires.
+	txnYieldBasisPoints := uint64(0)
+	txnYieldEarningDurationNanoSecs := int64(0)
+	if leftYieldCurvePoint.LockupDurationNanoSecs < lockupDurationNanoSeconds {
+		txnYieldBasisPoints = leftYieldCurvePoint.LockupYieldAPYBasisPoints
+		txnYieldEarningDurationNanoSecs = leftYieldCurvePoint.LockupDurationNanoSecs
+	}
+	if rightYieldCurvePoint.LockupDurationNanoSecs == lockupDurationNanoSeconds {
+		txnYieldBasisPoints = rightYieldCurvePoint.LockupYieldAPYBasisPoints
+		txnYieldEarningDurationNanoSecs = rightYieldCurvePoint.LockupDurationNanoSecs
+	}
+
+	// Convert variables to a consistent uint256 representation. This is to use them in SafeUint256 math.
+	txnYieldBasisPoints256 := uint256.NewInt().SetUint64(txnYieldBasisPoints)
+	txnYieldEarningDurationNanoSecs256 := uint256.NewInt().SetUint64(uint64(txnYieldEarningDurationNanoSecs))
+
+	// Compute the yield associated with this operation, checking to ensure there's no overflow.
+	yieldFromTxn, err :=
+		CalculateLockupYield(txMeta.LockupAmountBaseUnits, txnYieldBasisPoints256, txnYieldEarningDurationNanoSecs256)
+	if err != nil {
+		return 0, 0, nil, errors.Wrap(err, "_connectCoinLockup")
+	}
+
+	// We check that the minted yield does not cause an overflow in the transactor's balance.
+	// In the case of DeSo being locked up, we must check that the resulting amount is less than 2**64.
+	if uint256.NewInt().Sub(MaxUint256, yieldFromTxn).Lt(transactorBalanceNanos256) {
```

Comment:
Something I was thinking is that maybe we can delete a bunch of Isaac's old lockup code. Specifically, it seems like we can use your new LockedBalanceEntrys to delete:
* PrefixLockedStakeByValidatorAndStakerAndLockedAt and all its associated stuff
* _connectUnlockStake and all its helpers

The only issue I see with this is that Isaac's version relies on "epoch number" whereas this relies on timestamp. But maybe that's ok in the sense that we can "fudge" our unlocks using time rather than epochs, eg setting the unlock as "3 hours from now" rather than "3 epochs from the current epoch." This would introduce a risk that people could unlock a little sooner than they're supposed to. But I don't think this would be a real issue. And it would cause a simplification of the code that could be very nice.

You may need Sofonias's help with this. We can also merge this and then do that consolidation as a follow-up btw. That may be a better/simpler way to do it. You test that all of this is working as intended and then do the consolidation.